src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/IncludeFlagUtils.cs

```diff
@@ -53,15 +53,39 @@ internal static Dictionary<string, LibraryIncludeFlags> FlattenDependencyTypes(
             {
                 if (result.ContainsKey(dependency.Name))
                 {
-                    result[dependency.Name] = dependency.IncludeType;
+                    if (IsDependencyPruned(dependency, specFramework?.PackagesToPrune))
+                    {
+                        result[dependency.Name] = LibraryIncludeFlags.None;
+                    }
+                    else
+                    {
+                        result[dependency.Name] = dependency.IncludeType;
+                    }
                 }
                 else
                 {
-                    result.Add(dependency.Name, dependency.IncludeType);
+                    if (IsDependencyPruned(dependency, specFramework?.PackagesToPrune))
+                    {
+                        result[dependency.Name] = LibraryIncludeFlags.None;
+                    }
+                    else
+                    {
+                        result.Add(dependency.Name, dependency.IncludeType);
+                    }
```

Comment:
Consider refactoring to reduce duplication.

In its current form, the outer if checks whether the key exists and decides the behavior. The inner code duplicates the check of IsDependencyPruned. This can be refactored by separating the IsDependencyPruned determination from the conditional that checks if the key exists. It also removes a layer of nested ifs.

Example:
foreach (var dependency in directDependencies)
{
    var includeType = IsDependencyPruned(dependency, specFramework?.PackagesToPrune)
        ? LibraryIncludeFlags.None
        : dependency.IncludeType;

    if (result.ContainsKey(dependency.Name))
    {
        result[dependency.Name] = includeType;
    }
    else
    {
        result.Add(dependency.Name, includeType);
    }
}

Reply:
good point, fixed in new push.